File: flink-ml-python/pyflink/ml/feature/normalizer.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
import typing

from pyflink.ml.wrapper import JavaWithParams
from pyflink.ml.param import FloatParam, ParamValidators
from pyflink.ml.feature.common import JavaFeatureTransformer
from pyflink.ml.common.param import HasInputCol, HasOutputCol, Param


class _NormalizerParams(
    JavaWithParams,
    HasInputCol,
    HasOutputCol
):
    """
    Params for :class:`Normalizer`.
    """

    P: Param[float] = FloatParam(
        "p",
        "The p norm value.",
        2.0,
        ParamValidators.gt_eq(1.0))

    def __init__(self, java_params):
        super(_NormalizerParams, self).__init__(java_params)

    def set_p(self, value: float):
        return typing.cast(_NormalizerParams, self.set(self.P, value))

    def get_p(self) -> float:
        return self.get(self.P)

    @property
    def p(self):
        return self.get_p()


class Normalizer(JavaFeatureTransformer, _NormalizerParams):
    """
    A Transformer that normalizes a vector to have unit norm using the given p-norm.
    """

    def __init__(self, java_model=None):
        super(Normalizer, self).__init__(java_model)

    @classmethod
    def _java_transformer_package_name(cls) -> str:
        return "normalizer"

    @classmethod
    def _java_transformer_class_name(cls) -> str:
        return "Normalizer"

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/normalizer/Normalizer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.ml.feature.normalizer;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.typeutils.RowTypeInfo;
import org.apache.flink.ml.api.Transformer;
import org.apache.flink.ml.common.datastream.TableUtils;
import org.apache.flink.ml.linalg.BLAS;
import org.apache.flink.ml.linalg.IntDoubleVector;
import org.apache.flink.ml.linalg.typeinfo.VectorTypeInfo;
import org.apache.flink.ml.param.Param;
import org.apache.flink.ml.util.ParamUtils;
import org.apache.flink.ml.util.ReadWriteUtils;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.api.internal.TableImpl;
import org.apache.flink.types.Row;
import org.apache.flink.util.Preconditions;

import org.apache.commons.lang3.ArrayUtils;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/** A Transformer that normalizes a vector to have unit norm using the given p-norm. */
public class Normalizer implements Transformer<Normalizer>, NormalizerParams<Normalizer> {
    private final Map<Param<?>, Object> paramMap = new HashMap<>();

    public Normalizer() {
        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
    }

    @Override
    public Table[] transform(Table... inputs) {
        Preconditions.checkArgument(inputs.length == 1);
        StreamTableEnvironment tEnv =
                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());

        RowTypeInfo outputTypeInfo =
                new RowTypeInfo(
                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), VectorTypeInfo.INSTANCE),
                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));

        DataStream<Row> output =
                tEnv.toDataStream(inputs[0])
                        .map(new NormalizationFunction(getP(), getInputCol()), outputTypeInfo);

        Table outputTable = tEnv.fromDataStream(output);
        return new Table[] {outputTable};
    }

    @Override
    public void save(String path) throws IOException {
        ReadWriteUtils.saveMetadata(this, path);
    }

    public static Normalizer load(StreamTableEnvironment env, String path) throws IOException {
        return ReadWriteUtils.loadStageParam(path);
    }

    @Override
    public Map<Param<?>, Object> getParamMap() {
        return paramMap;
    }

    /** Normalization function that transforms a vector to have unit norm. */
    private static class NormalizationFunction implements MapFunction<Row, Row> {
        private final double p;
        private final String inputCol;

        public NormalizationFunction(double p, String inputCol) {
            this.p = p;
            this.inputCol = inputCol;
        }

        @Override
        public Row map(Row row) throws Exception {
            IntDoubleVector inputVec = row.getFieldAs(inputCol);
            IntDoubleVector outputVec = inputVec.clone();
            double norm = BLAS.norm(inputVec, p);
            BLAS.scal(1.0 / norm, outputVec);
            return Row.join(row, Row.of(outputVec));
        }
    }
}
